File: leetcode/leetcode_p398/code_398.py
# 力扣第398题，随机数索引
# 给定一个可能含有重复元素的整数数组，要求随机输出给定的数字的索引。可以假设数字一定存在与数组中。

import random


def picks(nums, target):
    index = []
    total = 0
    for i,v in enumerate(nums):
        if v == target:
            index.append(i)
            total += 1

    rand_index = random.randint(0, total-1)
    print(index[rand_index])
    return index[rand_index]


class Solution:
    def __init__(self, nums):
        self.nums = nums
        self.count = 0
        self.dict = {}

    def pick(self, target):
        if target not in self.dict:
            self.dict[target] = []
            for i, v in enumerate(self.nums):
                if v == target:
                    self.dict[target].append(i)
        self.count += 1
        index_list = self.dict[target]
        index = index_list[self.count % len(index_list)]
        return index


array = [1, 2, 3, 3, 3]
target = 3

p1 = picks(array, target)
print('p1: ', p1)

S1 = Solution(array)
p2 = S1.pick(target)
print('p2: ', p2)

File: leetcode/leetcode_p398/code_398.cpp
// 力扣第398题，随机数索引
// 给定一个可能含有重复元素的整数数组，要求随机输出给定的数字的索引，可以假设给定的数字一定存在与数组中。

#include <iostream>
#include <vector>
#include <unordered_map>
#include <string>
using namespace std;

class Solution{
public:
    vector<int> nums_vector;
    unordered_map<int, vector<int>> dict;
    int count = 0;
    Solution(vector<int>& nums){
        nums_vector = nums;
    }
    int pick(int target){
        int n = nums_vector.size();
        if (dict.find(target) == dict.end()){
            vector<int> temp;
            for (int i = 0; i < n; ++i)
            {
                if (target == nums_vector[i]){
                    temp.push_back(i);
                }
            }
            dict.emplace(target, temp);
        }
        count++;
        vector<int> index_list = dict.at(target);
        int rand = count % index_list.size();
        int index = index_list[rand];
        return index;
    }
};

int main()
{
    unordered_map<int, string> intstrMap;
    intstrMap.emplace(1, "num1");
    if (intstrMap.find(1) != intstrMap.end())
        cout << "intstrMap at 1: " << intstrMap.at(1) <<endl;
    if (intstrMap.find(0) == intstrMap.end())
        cout << "0 not in intstrMap!" << endl;
    
    vector<int> nums = {1, 2, 3, 3, 3, 4, 5};
    int target = 3;

    Solution * obj = new Solution(nums);
    int index0 = obj->pick(target);
    cout << "find target " << target << " in " << index0 << endl;

    cin.get();
    return 0;
}